### DISCUSSION TITLE
Sum of null values

### DISCUSSION THREAD

**Discussion by zueblin** [Kotlin/dataframe#936] [Q&A] [answered]
When calculating the sum of a column containing only null values, the result is 0. To me this seems wrong.
To reproduce:

```
val age_int = listOf(null, null).toColumnOf<Int?>("AGE")
println(dataFrameOf(age_int).sum("AGE"))
```
Output: 
```
0
```

The documentation states that the sum operation ignores null values. I would expect the result to be null instead of 0. 
Is this a bug, or does it work as designed? If its not a bug, would it be possible to control the behaviour in such a case? Or am i missing possible workarounds?

Thanks for you ideas!

**Comment by Jolanrensen** [comment, marked as answer]
It works as intended. The sum operation "ignoring" null values means it interprets them as the zero-value of the type you give it. So since you create a column of `Int?`, the zero value becomes `0`.

If we don't do this by default it's ambiguous what "summing nulls" means. This is also why in the standard library, there are no `.sum()` functions for nullable iterables, just `.sumOf {}` which expects you to replace nulls.

So, if you expect "all nulls" to become `null`, you could write something like:
`val sumOrNull: Int? = if (age_int.allNulls()) null else age_int.sum()`

or as extension function on a `DataColumn`:
```kt
fun <C : Number> DataColumn<C?>.sumOrNull(): C? =
    if (allNulls()) null else sum()
```

**Reply by zueblin** [reply]
Many thanks for the explanation and suggested workarounds. I gave the extension function a quick try and i think i will be able to solve my problem with this approach.
